proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,262 @@
+```
+<>:11:3: error: Opcode WriteSamplerFeedbackLevel not valid in shader model cs_6_4.
+note: at 'call void @dx.op.writeSamplerFeedbackLevel(i32 176, %dx.types.Handle %1, %dx.types.Handle %2, %dx.types.Handle %3, float %8, float %9, float undef, float undef, float %11)' in block '#0' of function 'main'.
+Validation failed.
+```
+
+This error displays the LLVM IR text for the error site. In a trivial example
+like the one above finding the error is easy. In a complex shader that could be
+thousands of lines long this is unmanageable.
+
+With this proposal under the _relaxed_ mode, clang will emit the following
+warning:
+
+```
+<>:11:6: warning: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   11 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+With this proposal under the _default_ or _strict_ mode, clang will emit the
+following error:
+
+```
+<>:11:6: error: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   11 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+To illustrate the difference between _default_ and _strict_ take the following
+example:
+
+```c++
+Texture2D texture : register(t0);
+SamplerState samplerState : register(s0);
+
+FeedbackTexture2D<SAMPLER_FEEDBACK_MIN_MIP> map : register(u0);
+
+void fn(uint3 threadId) {
+  float2 uv = threadId.xy;
+  uv /= 256;
+  map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+}
+
+[numthreads(4, 4, 1)] 
+void main(uint3 threadId : SV_DispatchThreadId) { }
+```
+
+In this example the call of the unavailable API is in an unused function. In
+both the _relaxed_ and _default_ diagnostic modes, clang will emit no
+diagnostics. In the _strict_ mode, clang emits the following diagnostic:
+
+```
+<>:9:6: error: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   9 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+     |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+In this case the DXIL validator will emit no diagnostic since the call to the
+unavailable API is not present in the final output.
+
+In all three cases above the DXIL validator will also emit the same diagnostic
+as it does today.
+
+To illustrate the behavior for library shaders consider the following example:
+
+> [Godbolt Link](https://godbolt.org/z/rKMGz566M)
+```c++
+float d(float f) {
+  return ddx(f);
+}
+
+float dead(float f) {
+  return WaveMultiPrefixSum(f, 1.xxxx);
+}
+
+float also_dead(float f) {
+  return ddy(f)
+}
+
+[shader("vertex")]
+float main() : FOO {
+  float f = 3;
+  return d(f);
+}
+```
+
+When compiled with the `lib_6_3` profile under the _default_ mode, clang will
+emit the following error:
+
+```
+<>:1:9: error: 'ddx' is available for pixel shaders beginning with Shader Model 2.0 
+   9 |   return ddx(f);
+     |          ^~~

```

Comment:
It might be more clear to say ```error: 'ddx' is not available in library shaders``` 

Do we have a way to encode intrinsics availability depending on the shader stage? The ddx/ddy are tricky as they are available in pixel shaders since 2.0+, but for compute, mesh and amplification shaders since 6.6+.

Reply:
We could add new clang attribute to annotate intrinsics availability per shader stage:

```c++
__attribute__((availability_shader_stage(pixel, introduced = 2.0)))
__attribute__((availability_shader_stage(compute, introduced = 6.6)))
float ddx(float);
```

Reply:
I agree that this is confusing.  If you use an intrinsic in a shader stage that is allowed in that shader stage, but in a newer shader model, then you should get a diagnostic that's relevant to the entry point type, not one that tells you about when it was first introduced elsewhere.

We also might need to add details about optional features required, especially if we want to move to a model where you must explicitly specify allowed features to prevent accidentally using optional features you didn't intend to use (which would be more consistent with SPIR-V targets).  That's directly relevant to this intrinsic because ddx is also usable in mesh and amplification shaders in SM 6.6+, but there it requires an optional feature bit to be set.

Reply:
I should mention that I don't agree with the error suggested above though.  I think it should say something more like: `error: 'ddx' is not available in vertex shaders`, since that's the entry function type that's violating the constraint.

If it were used in a compute shader, it should say something like: `error: 'ddx' is available for compute shaders beginning with Shader Model 6.6`.

Reply:
I think we can do this without adding a new attribute. The existing availability attribute has some sub-platform cases today for Apple's "App Extensions".

We can add `shadermodel_pixel` as a supported parameter to the existing attribute and get the support we need through one attribute.

Reply:
We are going to need to add at least `shadermodel_pixel`, `shadermodel_compute`, `shadermodel_mesh`, `shadermodel_amplification` just to cover just the derivate intrinsics. Let's see if we can get away with adding so many new values.

Reply:
> We are going to need to add at least `shadermodel_pixel`, `shadermodel_compute`, `shadermodel_mesh`, `shadermodel_amplification` just to cover just the derivate intrinsics. Let's see if we can get away with adding so many new values.

I don't think that will be a problem. This isn't really a significant maintenance burden to the compiler.

Reply:
The existing diagnostics in clang take this form:

```
<>:2:12: warning: 'WaveActiveCountBits' is only available on HLSL ShaderModel 6.0 or newer [-Wunguarded-availability-new]
    2 |     return hlsl::WaveActiveCountBits(b);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~
<>/hlsl/hlsl_intrinsics.h:1374:34: note: 'WaveActiveCountBits' has been marked as being introduced in HLSL ShaderModel 6.0 here, but the deployment target is HLSL ShaderModel 5.0
 1374 | __attribute__((convergent)) uint WaveActiveCountBits(bool Val);
      |    
```
[Compiler Explorer](https://godbolt.org/z/sqKfbbK9j) (ignore the unrelated error)

It is probably nice to keep the error text brief and elaborate in a note.

Reply:
> The example above was compiled with the lib_6_3 profile, not a vertex shader.
While that's true about the profile, the reason it's an error is because it's used in a vertex shader entry function in that library.

The closest thing to a "library shader" is an exported function in a library.  That would have a different message still, such as: `error: 'ddx' is not available for use in exported functions`.  The term "library shader" isn't really useful to refer to any shader in a library, since it's not specific enough.

Ideally, there should be no difference between what's allowed in a vertex shader entry function compiled to vs_6_4 and that same function compiled into a lib_6_4 library, thus the diagnostic language should also be the same, or as close as possible (if there's some justified difference).

Reply:
> We also might need to add details about optional features required, especially if we want to move to a model where you must explicitly specify allowed features to prevent accidentally using optional features you didn't intend to use (which would be more consistent with SPIR-V targets). That's directly relevant to this intrinsic because ddx is also usable in mesh and amplification shaders in SM 6.6+, but there it requires an optional feature bit to be set.

I would like to scope this proposal to HLSL intrincisc availability as related to shader model versions. We should have a separate discussion/proposal about availability diagnostics related to target runtimes (like Vulcan support @Keenuts mentioned below) and how to specify required optional features in HLSL.

Reply:
Filed an issue to track clarifying and reworking the diagnostic text:
https://github.com/microsoft/hlsl-specs/issues/188

Reply:
> I would like to scope this proposal to HLSL intrincisc availability as related to shader model versions. We should have a separate discussion/proposal about availability diagnostics related to target runtimes (like Vulcan support @Keenuts mentioned below) and how to specify required optional features in HLSL.

I don't agree with excluding Vulkan support.

I think we should have the same diagnostic infrastructure for all runtimes, and I think that is very reasonable without really expanding the design scope.

The only thing we should need to do is annotate the APIs with Vulkan availability in addition to Shader Model availability. I'm not suggesting that you we do that as part of this effort, but perhaps @keenuts can help identify and annotate the APIs we've already added in Clang.

Reply:
I think the remaining issues here are tracked in #188, so I’ll resolve the comment.